File: src/main/java/eu/qwsome/sql/Column.java
package eu.qwsome.sql;

import eu.qwsome.sql.condition.ValueHolder;


/**
 * This class represents a column in a database table.
 *
 * @author Lukáš Kvídera
 */
public class Column implements ValueHolder {

  private final String name;

  /**
   * Constructs column with a name.
   *
   * @param name of column
   */
  private Column(final String name) {
    this.name = name;
  }

  /**
   * Constructs column with a name.
   *
   * @param name of column
   * @return a {@link Column} instance
   */
  public static Column column(final String name) {
    return new Column(name);
  }

  /**
   * Constructs column with name and sourceTableAlias.
   *
   * @param name             of column
   * @param sourceTableAlias alias of table to which the column belongs
   */
  public static Column column(final String sourceTableAlias, final String name) {
    return new Column(sourceTableAlias.concat(".").concat(name));
  }

  @Override
  public String getSql() {
    return this.name;
  }

  @Override
  public String getValue() {
    return null;
  }

  @Override
  public ValueHolder apply(final String functionName) {
    return new FunctionWrappedHolder(functionName, this);
  }
}

File: src/main/kotlin/eu/qwsome/sql/Column.kt
/*
 * This class represents a column in a database table.
 *
 * @author Lukáš Kvídera
 */
package eu.qwsome.sql

import eu.qwsome.sql.condition.ValueHolder
import eu.qwsome.sql.condition.withFunction

class Column private constructor(private val name: String) : ValueHolder {

    override fun getSql(): String = name

    override fun getValue(): String? = null

    override fun apply(functionName: String): ValueHolder = withFunction(functionName)

    companion object {
        /**
         * Constructs column with a name.
         *
         * @param name of column
         * @return a [Column] instance
         */
        @JvmStatic
        fun column(name: String) = Column(name)

        /**
         * Constructs column with name and sourceTableAlias.
         *
         * @param sourceTableAlias alias of table to which the column belongs
         * @param name of column
         */
        @JvmStatic
        fun column(sourceTableAlias: String, name: String) = Column("$sourceTableAlias.$name")
    }
}
